src/main/scala/is/hail/variant/Call.scala

```diff
@@ -3,113 +3,318 @@ package is.hail.variant
+
+  def isDiploid(c: Call): Boolean = ploidy(c) == 2
+
+  def isUnphasedDiploid(c: Call): Boolean = (c & 0x7) == 4
+
+  def isPhasedDiploid(c: Call): Boolean = (c & 0x7) == 5
+
+  def ploidy(c: Call): Int = (c >>> 1) & 0x3
+
+  def alleleRepr(c: Call): Int = c >>> 3
+
+  def allelePair(c: Call): AllelePair = {
+    require(isDiploid(c), s"invalid ploidy: ${ ploidy(c) }. Only support ploidy == 2")
+
+    if (isPhased(c)) {
+        val p = Genotype.allelePair(alleleRepr(c))
+        AllelePair(p.j, p.k - p.j)
+    } else
+      Genotype.allelePair(alleleRepr(c))
+  }
+
+  def unphasedDiploidGtIndex(c: Call): Int = {
+    require(isUnphasedDiploid(c), s"Only support ploidy == 2 and unphased. Found ${ Call.toString(c) }.")
+    alleleRepr(c)
+  }
+
+  def alleles(c: Call): Array[Int] = {
+    (ploidy(c): @switch) match {
+      case 0 => Array.empty[Int]
+      case 1 => Array(alleleByIndex(c, 0))
+      case 2 => allelePair(c).alleleIndices
+      case _ => throw new UnsupportedOperationException
     }
+  }
 
-  def check(call: Call, nAlleles: Int) {
-    val nGenotypes = triangle(nAlleles)
-    assert(call >= 0 && call < nGenotypes, s"Invalid genotype found `$call' for number of alleles equal to `$nAlleles'.")
+  def alleleByIndex(c: Call, i: Int): Int = {
+    (ploidy(c): @switch) match {
+      case 0 => throw new UnsupportedOperationException
+      case 1 =>
+        require(i == 0)
+        alleleRepr(c)
+      case 2 =>
+        require (i == 0 || i == 1)
+        val p = allelePair(c)
+        if (i == 0) p.j else p.k
+      case _ =>
+        require(i >= 0 && i < ploidy(c))
+        alleles(c)(i)
+    }
   }
 
-  def genArb: Gen[Call] =
-    for {
-      v <- Variant.gen
-      nAlleles = v.nAlleles
-      nGenotypes = triangle(nAlleles)
-      call <- Gen.choose(0, nGenotypes - 1)
-    } yield {
-      check(call, nAlleles)
-      call
+  def parse(s: String): Call = Parser.parseCall(s)
+
+  def toString(c: Call): String = {
+    val phased = isPhased(c)
+    val sep = if (phased) "|" else "/"
+
+    (ploidy(c): @switch) match {
+      case 0 => if (phased) "|-" else "-"
+      case 1 =>
+        val a = alleleByIndex(c, 0)
+        if (phased) s"|$a" else s"$a"
+      case 2 =>
+        val p = allelePair(c)
+        s"${ p.j }$sep${ p.k }"
+      case _ =>
+        alleles(c).mkString(sep)
     }
+  }
+
+  def vcfString(c: Call): String = {
+    val phased = isPhased(c)
+    val sep = if (phased) "|" else "/"
 
-  def genNonmissingValue: Gen[Call] =
-    for {
-      v <- Variant.gen
-      nAlleles = v.nAlleles
-      nGenotypes = triangle(nAlleles)
-      call <- Gen.choose(0, nGenotypes - 1)
-    } yield {
-      check(call, nAlleles)
-      call
+    (ploidy(c): @switch) match {
+      case 0 => throw new UnsupportedOperationException("VCF spec does not support 0-ploid calls.")
+      case 1 =>
+        if (phased)
+          throw new UnsupportedOperationException("VCF spec does not support phased haploid calls.")
```

Comment:
really? Isn't a haploid call always phased?

Reply:
or never phased, I guess.

Reply:
anyway, I think we should be able to export it.

Reply:
Our Call rep now supports ploidy 0-2 (with 3-infty coming) and an independent phase bit.  That means haploid can be phased or unphased.

VCF doesn't have this distinction (we're treating VCF haploid as unphased), so rather than having export_vcf be lossy, we thought we'd throw an error on the cases the cases that VCF cannot represent.  This seems like a safe principle (in the same way, say, export_plink errors out on multi-allelic variants).

Phasing of haploid is complicated.  It is always phased with other haploid calls on the same strand, but what about with respect to PAR diploid calls?  Neither us nor VCF have a way to represent the latter (maybe you want something like m/1 vs 1/m to indicate a haploid call with position information?)  Anyway, we made the decision not to handle this case now since it is a big(ger) change to the call interface.